Store newly send value in internal variable inside CurrentValueObject

FranzBusch opened the pull request:
This should fix #38 but I think it breaks the synchronisation test. I tried fixing it and from what I can see everything is locked properly but apparently I am missing something since that test is flaky for me locally. Would be great if you can have a look and tell me what you think about this implementation.

FranzBusch committed:
Store newly send value in internal variable inside CurrentValueObject

broadwaylamb reviewed (approved):
This looks amazing, thank you!

I've restarted the tests, they can be a little flaky indeed.

And can you please address the [SwiftLint warnings](https://travis-ci.org/broadwaylamb/OpenCombine/jobs/565954391)?

MaxDesiatov commented:
This is slightly off-topic, but I have found that https://www.houndci.com/ (in case you didn't try it yet) works perfectly well for SwiftLint on GitHub projects: it's able to run SwiftLint on its own servers and it posts discovered errors and warnings as PR review comments. I'm not affiliated with that project in any way, just mentioning this because it has saved a lot of time during code review for me. You'd only need to register there (free for public repos) and add `.hound.yml` file with this content to specify your custom SwiftLint rules:

```yml
swiftlint:
  config_file: .swiftlint.yml
```

broadwaylamb commented:
@MaxDesiatov I'll definitely check this out, thanks!

broadwaylamb commented on Tests/OpenCombineTests/CurrentValueSubjectTests.swift:
```diff
@@ -310,6 +310,39 @@ final class CurrentValueSubjectTests: XCTestCase {
+
+        // When
+        passthrough.send(2)
+        print(passthrough.value)
```
Can you please remove this print?

broadwaylamb commented on Tests/OpenCombineTests/CurrentValueSubjectTests.swift:
```diff
@@ -310,6 +310,39 @@ final class CurrentValueSubjectTests: XCTestCase {
+
+        // When
+        passthrough.value = 3
+        print(passthrough.value)
```
And this too

broadwaylamb commented on Sources/OpenCombine/CurrentValueSubject.swift:
```diff
@@ -14,20 +14,33 @@ public final class CurrentValueSubject<Output, Failure: Error>: Subject {
+            defer {
+                _lock._unlock()
+            }
+            _lock._lock()
```
I see that the synchronization test failure reproduces every time. Have you tried removing this locking?

FranzBusch committed:
Fix code review and tests

FranzBusch commented:
I removed the print statements and fixed the SwiftLint warning. @MaxDesiatov suggestion for hound is I think a really good one, we used that in our projects before as well. Another solution could be to setup Danger on the CI with that you can also enforce many more conventions that you want to have in your PRs.

Lastly, regarding the test failure. I banged my head a bit against that and then realised it is really easy. The `+=/-=` operation is not thread safe and we cannot synchronise it from within the `CurrentValueSubject`. I quickly verified it against Apples implementation and also there I get synchronisation problems. To fix the test I locked around the += operation. @broadwaylamb What do you think about this?

FranzBusch committed:
Merge branch 'master' into bugfix/currentvaluesubject-send

broadwaylamb commented on Sources/OpenCombine/CurrentValueSubject.swift:
```diff
@@ -22,16 +24,23 @@ public final class CurrentValueSubject<Output, Failure: Error>: Subject {
+            defer {
+                _lock._unlock()
+            }
+            _lock._lock()
```
I'm not sure Combine uses synchronized access to the `value` property though. Maybe that's why the test failed.

FranzBusch replied:
I would guess that they at synchronise the access to value otherwise this would not be thread safe at all. The test failed because what tested is not something the the Combine/OpenCombine framework can synchronise but rather the user of the framework needs to synchronise.

broadwaylamb replied:
I've just checked the assembly, access to `value` is indeed not thread-safe.

FranzBusch replied:
Hm, I tested the value in access in Combine and at least the thread sanitiser doesn't complain. What do you want me to do here?

FranzBusch replied:
I removed the locks from within the get and set and the sanitiser didn't complain. So I removed it now.

broadwaylamb replied:
In this case, I think I'm ready to merge it, just waiting for you to push

broadwaylamb commented on Tests/OpenCombineTests/CurrentValueSubjectTests.swift:
```diff
@@ -441,13 +478,16 @@ final class CurrentValueSubjectTests: XCTestCase {
             {
+                lock.lock()
                 cvs.value -= 1
+                lock.unlock()
```
Maybe we should replace `cvs.value += 1` with just `cvs.send(42)`, and also remove this locking? It feels like it makes this test useless, since the access is guaranteed to be synchronized at test site, and we want to test the synchronization of `CurrentValueSubject` itself.

FranzBusch replied:
I removed the lock and just send 42 now. The thread sanitiser should catch if we have a sync problem here I guess.

broadwaylamb replied:
Thanks! Does the sanitizer complain if you remove the lock from the `value` getter and setter?

broadwaylamb commented:
@FranzBusch also can you please fix the conflicts so it can be merged?

FranzBusch committed:
remove lock from tests

FranzBusch commented:
@broadwaylamb Conflicts are resolved.

FranzBusch committed:
Remove lock from CurrentValueSubject getter

FranzBusch commented:
@broadwaylamb Done go ahead :)

broadwaylamb commented:
@FranzBusch will you remove that lock?

FranzBusch commented:
@broadwaylamb Ah sorry, though I pushed. Now :)

broadwaylamb commented on Sources/OpenCombine/CurrentValueSubject.swift:
```diff
@@ -22,16 +24,19 @@ public final class CurrentValueSubject<Output, Failure: Error>: Subject {
+            return _value
+        }
+        set {
+            send(newValue)
```
The backing is not set here. Maybe, since we don’t synchronize anymore, we can revert this to just a stored property with a `didSet` like it was before?
